feat: add theme presets

Cetttok opened the pull request:
created new module (themepresetmanager) and integrate it to conky in main.c.
implemented: 
- user can specify theme preset by alias in conifg option eg conky -c lean
- theme presets stored in git repo with simple struct (themealias.theme - folder name and everywords.conf - main config name)
- operations with git realise with libgit2 and I dont now there correct add includes ( in CMakeLists.txt files) 
- i dont know where i must to store my h and cpp files
- this code testes with /var/lib/conky/themes repo path and exapmle repo https://github.com/Cetttok/testRepoForConkyThemes
- this code cant clone repo with submodules and i dont know is it necessary
please view and say me what i must correct
i ready to perform yours requirements

Related to #2142

sorry for my English

Cetttok committed:
theme presets devel

Cetttok committed:
cmake add gitlib2

Cetttok committed:
cmake fixes

Cetttok committed:
requirements-dev.txt update

Cetttok committed:
requirements-dev.txt update 2

Cetttok committed:
docker scripts update

Cetttok committed:
docker scripts update 2

Cetttok committed:
Revert "docker scripts update 2"

This reverts commit 3ae0e8e279cd8de61e6bc5eb9a9079a3ea87c379.

Cetttok committed:
Revert "docker scripts update"

This reverts commit 4c97baadf24175cecd12e7d0cef7507e0d98d225.

Cetttok committed:
Revert "requirements-dev.txt update 2"

This reverts commit 688d5e2205c1b4524ecf85fdca2db43fe8adfc59.

Cetttok committed:
Revert "requirements-dev.txt update"

This reverts commit 8e04ed9956b06c5b3cf2e39ff142a0ba88078a33.

Cetttok committed:
docker scripts workflows update

Cetttok committed:
cmake update for testting

Cetttok committed:
oh it is python req

Cetttok committed:
macos libgit2 add

Cetttok committed:
try to fix nix

Caellian commented on CMakeLists.txt:
```diff
@@ -53,13 +53,16 @@ set(conky_sources ${CMAKE_BINARY_DIR}/config.h ${CMAKE_BINARY_DIR}/build.h)
 set(conky_libs ${conky_libs} toluapp_lib_static)
 set(conky_libs ${conky_libs} Vc)
+set(conky_libs ${conky_libs} PkgConfig::LIBGIT2)
+
```
Lua and Vc are added here because they're in thirdparty directory, but all the dependencies pulled from build system are added to `conky_libs` from `cmake/ConkyPlatformChecks.cmake` file. Search for `BUILD_IMLIB2` in CMake files for an example how to add libgit dependency.

Caellian commented on CMakeLists.txt:
```diff
@@ -74,6 +77,7 @@ if(BUILD_TESTING)
   include(Catch)
 endif()
 
+
```
Go through the diff, either here on github in "Files changed" tab or using `git diff` and remove these unrelated changes. This space for instace, blank comments later, etc. make the PR harder to review than it needs to be.

Caellian commented on src/conky.cc:
```diff
@@ -160,6 +160,7 @@
 
 #ifdef BUILD_BUILTIN_CONFIG
 #include "defconfig.h"
+#include "themespresetmanager.h"
```
Once you added a guard like `BUILD_IMLIB2`, the include needs to be guarded with `ifdef`s, as well as any code that's using that.

Caellian commented on src/main.cc:
```diff
@@ -300,7 +301,8 @@ inline void reset_optind() {
 }
-
+const char * THEME_PRESETS_REPO_CLONING_URL = "https://github.com/Cetttok/testRepoForConkyThemes"; // it is example need to create norm repo
+const char * THEME_PRESETS_REPO_PATH = "/var/lib/conky/themes";
```
Why are these in main.cc? You're declaring them here, just to pass them to the function in another file. Why not declare them above the function so they're easy to find and modify?

main.cc is already too large (much like conky.cc). Avoid placing more stuff in these two files.

Caellian commented on src/main.cc:
```diff
@@ -361,6 +367,10 @@ int main(int argc, char **argv) {
+      // case 'S':
+      //   std::cout << "optarg " <<  optarg << std::endl;
+      //   std::cout << presets.getThemePath(std::string(optarg)) << std::endl;
+      //   break;
```
Before merge, these comments need to be removed if the code isn't used. Very rarely it makes sense to leave code commented out - there's commit history for viewing old code if you need it. But let's keep that for final cleanup.

Caellian commented on src/main.cc:
```diff
@@ -389,10 +399,24 @@ int main(int argc, char **argv) {
   try {
     set_current_config();
-
+    tryToReplaceAliasToPresetPath(THEME_PRESETS_REPO_PATH, THEME_PRESETS_REPO_CLONING_URL );
```
`set_current_config` resolves paths to configs and _then_ you replace theme names with paths?

Shouldn't `tryToReplaceAliasToPresetPath` logic go into `set_current_config`, and before we do path resolution?

Ideally, there should be some easily recognizable disambiguation between a system path and a theme/preset name. Something like `theme:herpy_derpy_der`, so that you can easily and cheaply assume it's a theme, and not a directory relative to cwd (without asking the FS).

Caellian commented on src/themespresetmanager.cpp:
```diff
@@ -0,0 +1,288 @@
+#include <git2.h>
+#include <git2/common.h>
+const char * THEME_END_NAME = ".theme"; // can be changed at another one
+const char * THEME_CONFIG_END = ".conf"; // can be changed at another one
```
No need for these comments. It's FOSS, of course it can be changed :)

Caellian commented on src/themespresetmanager.cpp:
```diff
@@ -0,0 +1,288 @@
+    return false;
+  }
+  git_reference * mainBranch = nullptr;
+  if (git_reference_lookup(&mainBranch, _repo, "refs/remotes/origin/main")!=0){
```
Hardcoding "main" ref for repositories feels bad. It should be specified in the same place the repository URI is.

Caellian commented on src/themespresetmanager.h:
```diff
@@ -0,0 +1,95 @@
+  bool isGettedPathGitRepo(std::string pathToRepo);
+  git_repository * _repo = NULL;
+  // /std::string _pathToRepo = "";
+  std::string _repoUrl = "";
```
Don't prefix with `_` in C++. All `_` prefixed names are reserved for compiler and standard library implementations. See this [SO answer](https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier).

Caellian reviewed (changes requested):
Here's a preliminary review.

I'm not 100% sure about adding this feature and would like @brndnmtthws input before accepting this PR. It's a neat idea and I see how people would find it useful and convenient, but it also makes it very easy to distribute malicious configs because people won't check the presets/themes before updating them, just like they don't for AUR lol.

We could author an official repo with reviewed configs, but that's also a lot of work or trust to put into someone. So this would mostly be useful to Linux distribution authors.

Cetttok committed:
Major changes
- xml repo config
- git submodules support (i fuck that make)
- build optionts add
- code style change
- not full pull
- pulling one dir or file
- another small fixes

Cetttok committed:
libgit2 memory leaks and clang format format headers
- libgit2 memory leaks fixed
- clang-format on all themepresetsmanager module code
- increase verbosity of headers
- example xml added
- presets repo pulling/cloning log

Cetttok committed:
file names (theme-presets-manager module) changed

Cetttok committed:
blank spaces remove

Cetttok committed:
blank commit for restart test (nix macos failed?)

Cetttok commented:
Halo guys! I make most of all that  @Caellian told me.
That changed in last commits:
- repo not pulling of every path replacing request. only pulling one folder or subrepo of requested theme
- fixed memory leaks for libgit2 
- build options (bool BUILD_PRESETS string PRESETS_REPO_URL PRESETS_REPO_PATH)
- code used libgit masked in it option
- repo configuration file in xml (repo.conf) example in theme-presets-manager.h after advanced_repo_source declaration
- code in conky code style
- clang format on my module
- add loging about long operations (eg pulling ont theme) and theme presets repo that used
- adding support custom branches in non main repo (in submodules etc)
- git fetch git clone and another standart (and not standart) methods realised
thats not all.

Cetttok committed:
Revert "blank commit for restart test (nix macos failed?)"

This reverts commit 8f0a0f45b8d752771feea7352bd8916aca1062f5.

Cetttok committed:
Revert "blank spaces remove"

This reverts commit 259b01f54925469d4fc490ac572b1602f24ad81d.

Cetttok commented:
@Caellian i know that it is opensource (free of all)  but will be nice (for me) if you maybe say my Judgment for this

Caellian commented:
Thanks again for working on this. I want to give a clearer status update here, because I have intentionally not closed this PR.

I do think this is a useful direction for Conky. The best version of Conky definitely has a better story around discoverable, installable, and shareable configurations/themes. This PR is one of the few concrete attempts in that direction, so I do not want to discard it just because it is not immediately mergeable. It's also good work on your part: showcases good ability to navigate conky's large code-base, conky's CMake build system, and Linux environment well.

The reason I have not moved forward with it yet is that this is effectively a new subsystem, not a small feature. It introduces questions around dependency size, repository format, update/fetch behavior, default hosting, decentralization, trust, maintenance, testing, and security. Those decisions are _difficult to change later_ once users and distributions start relying on them.

There are a few things I would still want to rethink before merging something like this:

- I would like to minimize hard dependencies. I am not yet convinced Conky itself needs to depend on libgit2 for this. Static hosting, plain archives, or a simpler repository format may be enough.
- I would prefer the design to be decentralized, while still allowing us to point users toward a trustworthy default source (e.g. Ubuntu package repos).
  - I do not want users to be locked into one official repository in a way that requires rebuilding or patching the binary to use another source.
- I would like to avoid making XML a hard requirement.
- I think this should probably be tied to better visual/regression testing infrastructure, so that official/example configs can be tested in CI instead of silently breaking over time.

The largest concern, though, is security. Conky configs can execute Lua and shell commands, so a theme/config distribution mechanism can very easily become a malware distribution mechanism if we get the trust model wrong. This is especially risky for newer Linux users who may not understand that downloading and running a Conky theme is not equivalent to installing a harmless wallpaper.

Existing tools such as Conky Manager 2 handle this partly by having a much more controlled contribution path. That has downsides I don't want to see in a feature like this, but it also limits the risk of arbitrary third-party content being presented as safe or official. Conky has a very small maintainer/reviewer base (~2 people), so we need to be careful before adding an official-looking channel for downloading executable configurations. This adds a requirement to the infrastructure around it to amortize, at times, a slow review cycle.

So the current status is: I am keeping this open because I do not consider the idea rejected. An open PR is still there and visible. But I do not think now is the right time to merge it as-is. Before that, I think we need a clearer design for the repository model, trust boundaries, default source, dependency strategy, and testing story. I don't expect you to solve all of these on your own, and I'm letting the concept evolve in my head as well.

I may come back to this later after some of the surrounding infrastructure is in better shape. Even if this eventually needs to be reworked heavily, I still think the PR is valuable because it exposes the right set of problems and gives us something concrete to reason about. Once that happens, I will update your commit and keep you as author, ping you so you can chime in on it, and add any extras around it that might be needed.

tl;dr: this is valuable and aligned with where the project should go, but merging it now would commit us to infrastructure and policy we have not designed yet.

Note: I used an LLM to summarize my Ted talk long response.
